File: ver-fortran/src/quadtree.f90
module m_QuadTree
  implicit none
  type :: Node
    integer :: interaction_set(27)
    integer :: level
    integer :: id
    integer,allocatable :: vor_id(:)
    integer :: num_vor
    double precision :: center_pos(2)
    double precision :: region ! |x_min| = |x_max| = |y_min| = |y_max|
    complex(kind(0d0)),allocatable :: outer_coeffs(:)
    complex(kind(0d0)),allocatable :: inner_coeffs(:)
    logical :: is_leaf
  end type Node

  type :: QuadTree
    type(Node),allocatable :: nodes(:)! => null()
    integer :: nterms
    integer :: max_level
    double precision :: theta ! tolerance parameter
    integer :: level_hist(15) ! for debug
    integer :: t_level ! for biot_FMM
    double precision :: r_min ! for Tree
    integer :: stat ! for maxLevel loop
  end type QuadTree

  type :: Vortex
    double precision :: pos(2)
    double precision :: cir 
    double precision :: cor
  end type Vortex

  ! interface Quadtree
  !   module procedure init_QuadTree
  ! end interface Quadtree


contains
  !-------------------------------------------------------
  !- constructor for QuadTree
  !-------------------------------------------------------
  type(QuadTree) function init_QuadTree(max_level)
    integer,intent(in) :: max_level
    integer :: num_nodes
    type(Node),allocatable :: nodes(:)
    num_nodes = (4**(max_level+1)-1)/3 
    !print *,"-   initialize QuadTree, num_nodes:",num_nodes
    ! ↓レベルあげるとクッソ重くなる
    allocate(nodes(num_nodes))
    nodes(1)%level = 0
    nodes(:)%is_leaf = .false.
    nodes(:)%num_vor = 0
    init_QuadTree%nodes = nodes
    init_QuadTree%level_hist(:) = 0
  end function init_QuadTree
  !-------------------------------------------------------

  ! Othre subroutine
  function  quadrant_pos(pos,c_pos) result(quad_index)
    double precision,intent(in) :: pos(2) 
    double precision,intent(in) :: c_pos(2)
    integer :: i, quad_index
    quad_index = 0
    if (pos(1) >= c_pos(1)) then
      quad_index = ibset(quad_index,0)
    else 
      quad_index = ibclr(quad_index,0)
    end if
    if (pos(2) >= c_pos(2)) then
      quad_index = ibclr(quad_index,1)
    else 
      quad_index = ibset(quad_index,1)
    end if
    quad_index = quad_index + 1
    !print *,quad_index,pos
  end function quadrant_pos

  subroutine make_tree(vor,qtree,num_p,nterms,theta,t_level,r_min)
    type(Vortex),intent(in) :: vor(:)
    type(QuadTree), intent(out) :: qtree
    integer,intent(in) :: num_p
    integer,intent(in) :: nterms
    double precision,intent(in) :: theta
    integer,intent(in) :: t_level
    double precision,intent(in) :: r_min
    !type(QuadTree),target :: qtree


    ! integer,intent(in) :: max_level
    integer :: i
    integer :: num_nodes
    type(Node),allocatable :: nodes(:)
    integer,save :: max_level = 1

    !print *,"- make_tree"
    ! ----------------------------------
    ! - constructor
    ! ----------------------------------
    qtree%stat = 1 
    do while (qtree%stat == 1)
      qtree%max_level = max_level

      num_nodes = (4**(max_level+1)-1)/3 
      !print *,"-   initialize QuadTree, num_nodes:",num_nodes
      ! ↓レベルあげるとクッソ重くなる
      ! allocate(nodes(num_nodes))
      allocate(qtree%nodes(num_nodes))
      qtree%nodes(1)%level = 0
      qtree%nodes(:)%is_leaf = .false.
      qtree%nodes(:)%num_vor = 0
      ! qtree%nodes = nodes
      qtree%level_hist(:) = 0
      ! ----------------------------------

      ! qtree = QuadTree(max_level)
      qtree%nterms = nterms
      qtree%theta = theta
      qtree%t_level = t_level
      qtree%r_min = r_min 
      !p_qtree => qtree

      qtree%nodes(1)%num_vor = size(vor)
      qtree%nodes(1)%level = 0
      do i = 1,size(qtree%nodes)
        allocate(qtree%nodes(i)%inner_coeffs(qtree%nterms))
        allocate(qtree%nodes(i)%outer_coeffs(qtree%nterms))
        qtree%nodes(i)%outer_coeffs = cmplx(0.d0,0.d0)
        qtree%nodes(i)%inner_coeffs = cmplx(0.d0,0.d0)
      end do

      call add_vortex(vor,qtree)
      !call split_cell(vor,qtree,num_p)
      qtree%stat = 0
      call split_cell(vor,qtree, num_p, 1, .true.)
      if (qtree%stat == -1) then
        deallocate(qtree%nodes)
        qtree%stat = 1
        max_level = max_level + 1
      end if
    end do
    !call interaction_set(p_qtree)
  
  end subroutine make_tree
  
  
  subroutine add_vortex(vor,qtree)
    integer :: i 
    type(QuadTree), intent(inout) :: qtree
    type(Vortex), intent(in) :: vor(:)
    double precision :: region(2)
    double precision :: minValue
    !print *,"-   add_vortex"
    qtree%nodes(1)%num_vor = size(vor)
    allocate(qtree%nodes(1)%vor_id(size(vor)))
    do i = 1,size(vor)
      qtree%nodes(1)%vor_id(i) = i 
    end do
    !+0.1はセル格子とかぶった時にpos2indexが正しく計算されないので、少し広げる
    region(1) = abs(maxval(vor%pos(1)) - minval(vor%pos(1)))+0.0001
    region(2) = abs(maxval(vor%pos(2)) - minval(vor%pos(2)))+0.0001
    if (region(1) > region(2)) then
      qtree%nodes(1)%region = region(1)
    else 
      qtree%nodes(1)%region = region(2) 
    end if
    qtree%nodes(1)%center_pos(1) = 0.5*(maxval(vor(:)%pos(1)) + minval(vor(:)%pos(1)))
    qtree%nodes(1)%center_pos(2) = 0.5*(maxval(vor(:)%pos(2)) + minval(vor(:)%pos(2)))
  end subroutine add_vortex


  subroutine calc_CenterPos(quad_i,p_centerPos, region, c_centerPos)
    implicit none
    integer,intent(in) :: quad_i
    double precision,intent(in) :: p_centerPos(2)
    double precision,intent(in) :: region
    double precision,intent(out) :: c_centerPos(2)
    if (btest(quad_i-1,0)) then
      c_centerPos(1) = p_centerPos(1) + 0.25*region
    else 
      c_centerPos(1) = p_centerPos(1) - 0.25*region
    end if
    if (btest(quad_i-1,1)) then
      c_centerPos(2) = p_centerPos(2) - 0.25*region
    else 
      c_centerPos(2) = p_centerPos(2) + 0.25*region
    end if
  end subroutine calc_centerPos

  
  recursive subroutine split_cell(vor, qtree, num_p, p_index,flag)
    ! cf_ / child_first, p_ / parent
    use m_morton_order
    implicit none
    type(Vortex), intent(in) :: vor(:)
    type(QuadTree), intent(inout),target :: qtree
    integer,intent(in) :: num_p
    integer,intent(in) :: p_index
    logical,intent(in) :: flag
    integer :: c_index(4)
    integer :: i,j,k
    integer :: quad_index 
    integer :: p_num_vor
    integer :: node_index
    integer :: vor_index
    integer,allocatable :: p_quad(:,:)
    integer :: num_p_quad(4)
    type(Vortex) :: tmp_vor
    ! integer,allocatable :: vor_id(:)
    double precision :: x,y
    complex(kind(0d0)) :: z_j,z_star
    double precision,parameter :: pi = 2*acos(0.d0)
    type(Node),pointer :: c_cell


    ! quadrant loop
    if (qtree%nodes(p_index)%level + 1 > qtree%max_level) then
      qtree%stat = -1
      return 
      !print *,"Error: You cannot refer to levels higher than expected."
      !stop
    end if

    p_num_vor = qtree%nodes(p_index)%num_vor
    call getChild(p_index, c_index)
    do i = 1,4 
      allocate(qtree%nodes(c_index(i))%vor_id(p_num_vor))
      !qtree%nodes(c_index(i))%num_vor = 0
    end do

    ! vortex loop
    do i = 1,qtree%nodes(p_index)%num_vor
      tmp_vor = vor(qtree%nodes(p_index)%vor_id(i))
      quad_index = quadrant_pos(tmp_vor%pos,qtree%nodes(p_index)%center_pos)
      node_index = c_index(quad_index)
      qtree%nodes(node_index)%num_vor = qtree%nodes(node_index)%num_vor + 1
      qtree%nodes(node_index)%vor_id(qtree%nodes(node_index)%num_vor) = qtree%nodes(p_index)%vor_id(i)
    end do

    ! quadrant loop
    do i = 1, 4
      call  calc_centerPos(i, &
                           & qtree%nodes(p_index)%center_pos, &
                           & qtree%nodes(p_index)%region, &
                           & qtree%nodes(c_index(i))%center_pos)

      qtree%nodes(c_index(i))%id = c_index(i)
      qtree%nodes(c_index(i))%level = qtree%nodes(p_index)%level + 1
      qtree%nodes(c_index(i))%region = 0.5*qtree%nodes(p_index)%region
      if (qtree%nodes(c_index(i))%num_vor < num_p) then
        if (flag) then
          qtree%nodes(c_index(i))%is_leaf = .true.
          qtree%level_hist(qtree%nodes(c_index(i))%level) = &
              & qtree%level_hist(qtree%nodes(c_index(i))%level) + qtree%nodes(c_index(i))%num_vor
        end if
        if (qtree%nodes(c_index(i))%level < qtree%max_level) then
          ! already reached reaf-cell
          if (qtree%stat /= -1) call split_cell(vor, qtree, num_p, c_index(i), .false.)
        end if
      else
        ! not reach reaf-cell
        if (qtree%stat /= -1) call split_cell(vor, qtree, num_p, c_index(i), .true.)
      end if

    end do
    !print *,"-----------------------"

  end subroutine split_cell

end module m_QuadTree

File: ver-c/sample_implement/src-MPI/tree/quadtree.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h> 
#include "quadtree.h"
#include "morton_order.h"
#include "fmm.h"

QuadTree qtree;
Vortex *vors;

void read_fmm_param(){
  FILE *file;
  file = fopen("../cal_cond/fmm_param.dat","r");
  if(file==NULL){
    printf("Error: Could not read FMM parameters.");
    return;
  }

  fscanf(file,"%i    nterms",&qtree.nterms);
  fscanf(file,"%i   num_p",&qtree.num_p);
  fscanf(file,"%lf theta",&qtree.theta);  
  fscanf(file,"%lf  r_min ! for Tree",&qtree.r_min);
  
  fclose(file);
  return;
}

int quadrant_pos(double pos[2], double c_pos[2]){
  int quad_index;
  quad_index = 0;
  if (pos[0] >= c_pos[0]){ quad_index++; }
  if (pos[1] <= c_pos[1]){ quad_index += 2; }
  return quad_index;
}

double maxval(double a[],size_t n){
  int i;
  double  max;
  max = a[0];
  for(i=1; i<n; i++){
    if(a[i] > max){
      max = a[i];
    }
  }
  // for (i=0;i<10;i++){printf("aa:%f",a[i]);}
  return max;
}

double minval(double a[],size_t n){
  int i;
  double min;
  min = a[0];
  for(i=1; i<n; i++){
    if(a[i] < min){
      min = a[i];
    }        
  }
  return min;
}

void calc_CenterPos(int quad_i, double *p_centerPos, double region, double *c_centerPos){
  // printf("\n\n%i",quad_i);
  if (quad_i%2 != 0) {
    // printf("x+\n");
    c_centerPos[0] = p_centerPos[0] + 0.25*region;
  } else {
    // printf("x-\n");
    c_centerPos[0] = p_centerPos[0] - 0.25*region;
  }

  if (quad_i / 2 == 0 ){
    // printf("y+\n");
    c_centerPos[1] = p_centerPos[1] + 0.25*region;
  } else {
    // printf("y-\n");
    c_centerPos[1] = p_centerPos[1] - 0.25*region;
  }

}

void calc_region(double vor_b[][2], int nvor_b){
  int i;
  double values[nvor_b];
  double region[2];

  for (i=0; i<nvor_b; i++){ values[i] = vor_b[i][0]; }
  region[0] = fabs(maxval(values,nvor_b) - minval(values,nvor_b)) + 0.0001;
  for (i=0; i<nvor_b; i++){ values[i] = vor_b[i][1]; }
  region[1] = fabs(maxval(values,nvor_b) - minval(values,nvor_b)) + 0.0001;
  if (region[0] > region[1]) {
    qtree.node[0].region = region[0];
  } else {
    qtree.node[0].region = region[1];
  }
  for (i=0; i<nvor_b; i++){ values[i] = vor_b[i][0]; }
  qtree.node[0].center_pos[0] = 0.5*(maxval(values,nvor_b) + minval(values,nvor_b));
  for (i=0; i<nvor_b; i++){ values[i] = vor_b[i][1]; }
  qtree.node[0].center_pos[1] = 0.5*(maxval(values,nvor_b) + minval(values,nvor_b));
}


void make_tree(int nvor_b, 
               double vor_b[nvor_b][2],
               double cir[nvor_b], 
               double cor[nvor_b])
{
  int i;
  int num_nodes; // number of nodes
  static int max_level = 1;
  double region[2];

  // printf("make tree\n");
  vors= (Vortex *)malloc(sizeof(Vortex) * nvor_b);
  for (i=0; i<nvor_b; i++){
    vors[i].pos[0] = vor_b[i][0];
    vors[i].pos[1] = vor_b[i][1];
    vors[i].cir = cir[i];
    vors[i].cor = cor[i]; 
  }
  // qtree.nterms = 0;
  // qtree.theta = 0;
  // qtree.r_min = 0; 
  // qtree.num_p = 250;

  read_fmm_param();

  // qtree.nterms = nterms;
  // qtree.theta = theta;
  // qtree.r_min = r_min;

  // searching max_level and make tree
  // stat:  
  //   max_level is sufficient:     qtree.stat transitions from 1 to 0 and end the loop.
  //   max_level is not sufficient: After qtree.stat goes from 1 to 0, it goes to -1 in split_cell, detects it and goes back to 1 again.
  qtree.stat = 1;
  while (qtree.stat == 1){ 
    // printf("max_level: %i\n",max_level);
    for (i=0; i<15; i++) { qtree.level_hist[i] = 0; }
    qtree.max_level = max_level;

    // initialize node
    for (i=0; i<sizeof(qtree.node)/sizeof(qtree.node[0]); i++){
      if(qtree.node[i].outer_coeffs){ free(qtree.node[i].outer_coeffs); }
    }
    if (qtree.node) { free(qtree.node); }
    num_nodes = (pow(4,qtree.max_level+1)-1)/3;
    qtree.num_nodes = num_nodes;
    qtree.node = (Node*)malloc(sizeof(Node) * num_nodes);
    for (i=0; i<num_nodes; i++){
      qtree.node[i].is_leaf = 0;
      qtree.node[i].num_vor = 0;
      qtree.node[i].level = getLevel(i);
    }

    // qtree.node[0].interactionset = 0
    // initialize root cell
    qtree.node[0].level = 0;
    qtree.node[0].num_vor = nvor_b;
    qtree.node[0].outer_coeffs = (double complex *)malloc(sizeof(double complex) * qtree.nterms);
    // qtree.node[0].inner_coeffs = (double complex *)malloc(sizeof(double complex) * qtree.nterms);
    for (i=0; i<qtree.nterms; i++){ qtree.node[0].outer_coeffs[i] = 0; }

    // set vortex on root cell
    qtree.node[0].vor_id = (int*)malloc(sizeof(int) * nvor_b);
    for (i=0; i<nvor_b; i++){ qtree.node[0].vor_id[i] = i; }

    calc_region(vor_b, nvor_b);


    // split cell (recurcive function)
    qtree.stat = 0;
    split_cell(0,1);
    // printf("-----------\n");
    // for (i=0;i<sizeof(qtree.level_hist)/sizeof(qtree.level_hist[0]);i++){
    //   printf("level:%i\n",qtree.level_hist[i]);
    // }

    // Judgment of End
    if (qtree.stat == -1) {
      qtree.stat = 1;
      max_level++;
    }
  }

  S2M();
  M2M();




  // quadrant_pos(pos,c_pos);

}

// void calc_centerPos(){
// }

void split_cell(int p_index, int flag){
  int p_num_vor;
  int c_index[4];
  int i,j;
  int quad_index,node_index;
  Vortex tmp_vor;

  if (qtree.node[p_index].level + 1 > qtree.max_level){
    qtree.stat = -1;
    return;
  }

  p_num_vor = qtree.node[p_index].num_vor;
  
  // initialize child cell 
  getChild(p_index,c_index);
  for (i=0;i<4;i++){ 
    qtree.node[c_index[i]].vor_id = (int *)malloc(sizeof(int) * p_num_vor);
    for (j=0; j<p_num_vor; j++){
      qtree.node[c_index[i]].vor_id[j] = -1;
    }
  }

  // put vortex to child cell
  // printf("---------\n");
  for (i=0; i<p_num_vor; i++){
    tmp_vor = vors[qtree.node[p_index].vor_id[i]];
    quad_index = quadrant_pos(tmp_vor.pos,qtree.node[p_index].center_pos);
    node_index = c_index[quad_index];
    qtree.node[node_index].num_vor = qtree.node[node_index].num_vor + 1;
    qtree.node[node_index].vor_id[qtree.node[node_index].num_vor-1] = qtree.node[p_index].vor_id[i];
  }

  for (i=0;i<4;i++){

    calc_CenterPos(i, 
                      qtree.node[p_index].center_pos,
                      qtree.node[p_index].region, 
                      qtree.node[c_index[i]].center_pos);
    // printf("%i:%i\n",c_index[i],qtree.node[c_index[i]].num_vor);
    // printf("center:%f %f\n",qtree.node[c_index[i]].center_pos[0],qtree.node[c_index[i]].center_pos[1]);
    qtree.node[c_index[i]].id = c_index[i];
    qtree.node[c_index[i]].level = qtree.node[p_index].level + 1;
    qtree.node[c_index[i]].region = 0.5*qtree.node[p_index].region;
    qtree.node[c_index[i]].outer_coeffs = (double complex *)malloc(sizeof(double complex)*qtree.nterms);
    for (j=0; j<qtree.nterms; j++) { qtree.node[c_index[i]].outer_coeffs[j] = 0.0; }

    if (qtree.node[c_index[i]].num_vor < qtree.num_p) {
      if (flag) {
        qtree.node[c_index[i]].is_leaf = 1;
        qtree.level_hist[qtree.node[c_index[i]].level] += 
              qtree.node[c_index[i]].num_vor;
      }
      if (qtree.node[c_index[i]].level < qtree.max_level){
        if (qtree.stat != -1) { split_cell(c_index[i], 0); }
      }
    } else {
      if (qtree.stat != -1) { split_cell(c_index[i], 1); }
    }
  }

}
